lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -1062,22 +1075,96 @@ def get_service_status(  # noqa: C901
+class OpenSearchBackupFactory(OpenSearchPluginRelationsHandler):
+    """Creates the correct backup class and populates the appropriate relation details."""
 
-    This class is solely responsible for the creation of the correct S3 client manager.
+    _backup_obj = None
 
-    If this cluster is an orchestrator or failover cluster, then return the OpenSearchBackup.
-    Otherwise, return the OpenSearchNonOrchestratorBackup.
+    def __init__(self, charm: CharmBase):
+        super().__init__()
+        self._charm = charm
 
-    There is also the condition where the deployment description does not exist yet. In this
-    case, return the base class OpenSearchBackupBase. This class solely defers all s3-related
-    events until the deployment description is available and the actual S3 object is allocated.
-    """
-    if not charm.opensearch_peer_cm.deployment_desc():
-        # Temporary condition: we are waiting for CM to show up and define which type
-        # of cluster are we. Once we have that defined, then we will process.
-        return OpenSearchBackupBase(charm)
-    elif charm.opensearch_peer_cm.deployment_desc().typ == DeploymentType.MAIN_ORCHESTRATOR:
-        return OpenSearchBackup(charm)
-    return OpenSearchNonOrchestratorClusterBackup(charm)
+    @property
+    def _get_peer_rel_data(self) -> Optional[S3RelDataCredentials]:
+        """Returns the relation data.
+
+        If we have two connections here, then we check if data is the same. If not, then
+        return empty. Otherwise, return the credentials for s3 if available.
+        """
+        if not (relations := self._charm.model.relations.get(PeerClusterRelationName, [])):
+            # No relation currently available
+            return None
+
+        contents = [
+            data for rel in relations if (data := rel.data.get(rel.app, {}).get("data", None))
+        ]
+        if not contents or not all(contents):
+            # We have one of the peers missing data
+            return None
+
+        contents = [PeerClusterRelData.from_str(c) for c in contents]
+        if len(contents) > 1 and contents[0].credentials.s3 != contents[1].credentials.s3:
+            return None
+        return contents[0].credentials.s3
+
+    @override
+    def is_relation_set(self) -> bool:
+        """Checks if the relation is set for the plugin handler."""
+        if isinstance(self.backup(), OpenSearchBackup):
+            relation = self._charm.model.get_relation(self._relation_name)
+            return relation is not None and relation.units != {}
+
+        # We are not not the MAIN_ORCHESTRATOR
+        # The peer-cluster relation will always exist, so we must check if
+        # the relation has the information we need or not.
+        return (
+            self._get_peer_rel_data is not None
+            and self._get_peer_rel_data.access_key
+            and self._get_peer_rel_data.secret_key
```

Comment:
similarly, this should be secrets based

Reply:
So, I am using the relation to prototype the plugin with a relation handler